pydm/data_plugins/epics_plugins/pyepics_plugin_component.py

```diff
@@ -1,165 +1,98 @@
-                        # with Matt Newville
-                        self.new_value_signal[str].emit(char_value)
-                elif ftype in float_types:
-                    self.new_value_signal[float].emit(float(value))
-                else:
-                    self.new_value_signal[str].emit(char_value)
-
-    def update_ctrl_vars(self, units=None, enum_strs=None, severity=None, upper_ctrl_limit=None, lower_ctrl_limit=None, precision=None, *args, **kws):
-        if severity is not None and self._severity != severity:
-            self._severity = severity
-            self.new_severity_signal.emit(int(severity))
-        if precision is not None and self._precision != precision:
-            self._precision = precision
-            self.prec_signal.emit(precision)
-        if enum_strs is not None and self._enum_strs != enum_strs:
-            self._enum_strs = enum_strs
+        if value is not None:
+            self.data[DataKeys.VALUE] = value
+
+        self.send_to_channel()
+
+    def update_ctrl_vars(self, units=None, enum_strs=None, severity=None,
+                         upper_ctrl_limit=None, lower_ctrl_limit=None,
+                         precision=None, *args, **kws):
+        if severity is not None:
+            self.data[DataKeys.SEVERITY] = severity
+        if precision is not None:
+            self.data[DataKeys.PRECISION] = precision
+        if enum_strs is not None:
             try:
-                enum_strs = tuple(b.decode(encoding='ascii') for b in enum_strs)
+                enum_strs = tuple(
+                    b.decode(encoding='ascii') for b in enum_strs)
             except AttributeError:
                 pass
-            self.enum_strings_signal.emit(enum_strs)
-        if units is not None and len(units) > 0 and self._unit != units:
+            self.data[DataKeys.ENUM_STRINGS] = enum_strs
+        if units is not None and len(units) > 0:
             if type(units) == bytes:
                 units = units.decode()
-            self._unit = units
-            self.unit_signal.emit(units)
-        if upper_ctrl_limit is not None and self._upper_ctrl_limit != upper_ctrl_limit:
-            self._upper_ctrl_limit = upper_ctrl_limit
-            self.upper_ctrl_limit_signal.emit(upper_ctrl_limit)
-        if lower_ctrl_limit is not None and self._lower_ctrl_limit != lower_ctrl_limit:
-            self._lower_ctrl_limit = lower_ctrl_limit
-            self.lower_ctrl_limit_signal.emit(lower_ctrl_limit)
+            self.data[DataKeys.UNIT] = units
+        if upper_ctrl_limit is not None:
+            self.data[DataKeys.UPPER_LIMIT] = upper_ctrl_limit
+        if lower_ctrl_limit is not None:
+            self.data[DataKeys.LOWER_LIMIT] = lower_ctrl_limit
 
     def send_access_state(self, read_access, write_access, *args, **kws):
         if is_read_only():
-            self.write_access_signal.emit(False)
+            self.data[DataKeys.WRITE_ACCESS] = False
```

Comment:
Just a thought, but in `read_only` mode, should you even bother subscribing to the access changes?

Reply:
In theory one could switch the `read_only` flag during application execution...
One use case for that would be access control (not the EPICS one but a more advanced form) in which you could make the whole PyDM app go into a read only mode... 
Sounds reasonable?

Reply:
I see...

Seems a bit odd to me that `read_only` wouldn't be a read-only parameter set on startup, but OK otherwise.